fix(no-unused-vars): consider functions within conditional expressions usable

eventualbuddha opened the pull request:
Previously, a function declared directly within an `AstKind::ConditionalExpression` would be considered unused. This happened because `is_function_or_class_declaration_used` did not directly handle conditional expression parents, letting such cases fall into the default branch which just returned `false`, marking the function as unused. This change fixes that by considering such functions used when the conditional expression itself is used.

eventualbuddha committed:
fix(no-unused-vars): consider functions within conditional expressions usable

Previously, a function declared directly within an `AstKind::ConditionalExpression` would be considered unused. This happened because `is_function_or_class_declaration_used` did not directly handle conditional expression parents, letting such cases fall into the default branch which just returned `false`, marking the function as unused. This change fixes that by considering such functions used when the conditional expression itself is used.

eventualbuddha committed:
style(no-unused-vars): fix clippy lints

DonIsaac commented:
Seems like a pathological case but otherwise LGTM

DonIsaac commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/tests/eslint.rs:
```diff
@@ -364,6 +364,9 @@ fn test() {
             None,
         ),
         ("function* foo(cb) { cb = yield function(a) { cb(1 + a); }; } foo();", None), // { "ecmaVersion": 6 },
+        ("console.log(function a() {} ? b : c)", None), // { "ecmaVersion": 6 }
```
Can you add these to oxc test cases instead? These are the tests ported directly from ESLint

eventualbuddha replied:
a387d57

eventualbuddha commented:
> Seems like a pathological case but otherwise LGTM

This actually came up as a false positive in my [company's codebase](https://github.com/votingworks/vxsuite/blob/326edf3e3d1c7776d2ef571a6811d8042a92c2eb/libs/eslint-plugin-vx/src/rules/gts_no_default_exports.ts#L102-L103). We name the generator function there because our ESLint config enables the [`func-names` rule](https://eslint.org/docs/latest/rules/func-names).

eventualbuddha committed:
test(no-unused-vars): move to `oxc` from `eslint` suite

Per https://github.com/oxc-project/oxc/pull/6553#discussion_r1799496110, the `eslint` suite is for tests ported directly from ESLint.

camc314 commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/tests/oxc.rs:
```diff
@@ -54,6 +54,9 @@ fn test_vars_simple() {
             ",
             None,
         ),
+        ("console.log(function a() {} ? b : c)", None),
```
hopefully no one writes code like yhis. it'll always eval to b 🙂

DonIsaac committed:
Merge branch 'main' into fix/no-unused-vars/conditional-expression
